crates/stages/src/stages/index_storage_history.rs

```diff
@@ -0,0 +1,510 @@
+        let from_transition = tx.get_block_transition(stage_progress)?;
+
+        // NOTE: can probably done more probabilistic take of bundles with transition but it is
+        // guess game for later. Transitions better reflect amount of work.
+        let to_block =
+            std::cmp::min(stage_progress + self.commit_threshold, previous_stage_progress);
+        let to_transition = tx.get_block_transition(to_block)?;
+
+        tx.cursor_read::<tables::StorageChangeSet>()?
+            .walk((from_transition, Address::zero()).into())?
+            .take_while(|res| {
+                res.as_ref().map(|(k, _)| k.transition_id() < to_transition).unwrap_or_default()
+            })
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            // fold all storages to one set of changes
+            .fold(
+                BTreeMap::new(),
+                |mut storages: BTreeMap<(Address, H256), Vec<u64>>, (index, storage)| {
+                    storages
+                        .entry((index.address(), storage.key))
+                        .or_default()
+                        .push(index.transition_id());
+                    storages
+                },
+            )
+            .into_iter()
+            // insert indexes to StorageHistory.
+            .try_for_each(|((address, storage_key), mut indices)| -> Result<(), StageError> {
+                let mut last_shard = take_last_storage_shard(tx, address, storage_key)?;
+                last_shard.append(&mut indices);
+
+                // chunk indices and insert them in shards of N size.
+                let mut chunks = last_shard
+                    .iter()
+                    .chunks(NUM_OF_INDICES_IN_SHARD)
+                    .into_iter()
+                    .map(|chunks| chunks.map(|i| *i as usize).collect::<Vec<usize>>())
+                    .collect::<Vec<_>>();
+                let last_chunk = chunks.pop();
+
+                // chunk indices and insert them in shards of N size.
+                chunks.into_iter().try_for_each(|list| {
+                    tx.put::<tables::StorageHistory>(
+                        StorageShardedKey::new(
+                            address,
+                            storage_key,
+                            *list.last().expect("Chuck does not return empty list") as TransitionId,
+                        ),
+                        TransitionList::new(list).expect("Indices are presorted and not empty"),
+                    )
+                })?;
+                // Insert last list with u64::MAX
+                if let Some(last_list) = last_chunk {
+                    tx.put::<tables::StorageHistory>(
+                        StorageShardedKey::new(address, storage_key, u64::MAX),
+                        TransitionList::new(last_list)
+                            .expect("Indices are presorted and not empty"),
+                    )?;
+                }
```

Comment:
why the last list is inserted with `u64::MAX`? can we add more detailed description for `struct IndexStorageHistoryStage` explaining what it actually stores and how (with special cases like `u64::MAX`)?

Reply:
Wrote it here: https://github.com/paradigmxyz/reth/pull/978/files#diff-bc5df8783f831a8e01372f13a49fbd1d2179d21dc309c8fdba3604bcd0dd372dR223-R241